Add docs about multi cluster prom monitoring (#1490)
pingcap/docs-tidb-operator: zh/deploy-tidb-monitor-across-multiple-kubernetes.md

---

> title: 跨多个 Kubernetes 集群监控 TiDB 集群

Reviewer: add this doc into TOC?

Reviewer 2: 这里还要改下 https://docs.pingcap.com/zh/tidb-in-kubernetes/dev/deploy-tidb-cluster-across-multiple-kubernetes 这里，关于监控部署链接到这个文档

summary: 介绍如何对跨多个 Kubernetes 集群的 TiDB 集群进行监控，并集成到常见 Prometheus 多集群监控体系中
---

# 跨多个 Kubernetes 集群监控 TiDB 集群

本文档介绍如何对跨多个 Kubernetes 集群的 TiDB 集群进行监控，以及如何与几种社区常见的 Prometheus 多集群监控方式进行集成，从而实现统一全局视图进行监控数据访问。

## Push 方式

### 部署架构图

本文档以 Thanos 为例，如果使用了其他兼容 Prometheus Remote API 的中心化存储方案（如 Cortex、M3DB、Grafana Cloud 等），只需对 Thanos 相关组件进行替换即可。

![push-thanos-receive.png](/media/push-thanos-receive.png)

### 前置条件

多个 Kubernetes 集群间的组件满足以下条件：

- 各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件有能力访问 Thanos Receiver 组件。
- Grafana 组件有能力访问 Thanos Query 组件。

### 部署 TiDB 集群监控

> 根据不同 TiDB 集群所在的 Kubernetes 集群设置以下环境变量，其中`cluster_name`为TiDB集群名称，`cluster_namespace`为TiDB集群所在的命名空间，`kubernetes_cluster_name`为自定义的 kubernetes 集群名称，作为标识 Prometheus 的`externallabels`，`storageclass_name`设置为当前集群中的存储，`remote_write_url` 为`thanos-receiver`（或其他兼容 Prometheus remote API）组件的 host 信息，关于 thanos 部署方案参考 [kube-thanos](https://github.com/thanos-io/kube-thanos) 以及 [Example](https://github.com/pingcap/tidb-operator/tree/master/examples/monitor-prom-remotewrite)。

Reviewer 2: 英文、数字和中文之间加空格

```sh
cluster_name="cluster1"
cluster_namespace="pingcap"
kubernetes_cluster_name="kind-cluster-1"
storageclass_name="local-storage"
remote_write_url="http://thanos-receiver:19291/api/v1/receive"
```

执行以下指令，创建`TidbMonitor`：

> ```sh

Reviewer: 可以用一些 `{{< copyable "shell-regular" >}}` 指令

Reviewer 3: 文档站目前有效的高亮是 ` ```bash`。其他像` ```sh` 和 ` ```shell` 都不能渲染哈。
See https://github.com/pingcap/docs-tidb-operator/issues/1501

cat << EOF | kubectl apply -n ${cluster_namespace} -f -
apiVersion: pingcap.com/v1alpha1
kind: TidbMonitor
metadata:
  name: ${cluster_name}
spec:
  clusters:
  - name: ${cluster_name}
    namespace: ${cluster_namespace}
  externalLabels:

>     #kubernetes indicates the k8s cluster name, you can change the label's name on your own, but you should notice that `cluster` label has been used by tidb already. For more information, please refer to issue #4219.

Reviewer: 这个 `#4219` 要说明一下是 `tidb-operator` repo 吗？

Reviewer 3: 直接放链接吧

![pull-thanos-query.png](/media/pull-thanos-query.png)

#### 前置条件

需要配置 Kubernetes 的网络和 DNS，使得 Kubernetes 集群满足以下条件：

- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件的 Pod IP。
- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件的 Pod FQDN。

#### 部署 TiDB 集群监控

根据不同 TiDB 集群所在的 Kubernetes 集群设置以下环境变量，其中 `cluster_name`为TiDB集群名称，`cluster_namespace`为TiDB集群所在的命名空间，`kubernetes_cluster_name`为自定义的 kubernetes 集群名称，作为标识 Prometheus 的`externallabels`，`storageclass_name`设置为当前集群中的存储，关于 thanos 部署方案参考 [kube-thanos](https://github.com/thanos-io/kube-thanos) 以及 [Example](https://github.com/pingcap/tidb-operator/tree/master/examples/monitor-with-thanos)。

```sh
cluster_name="cluster1"
cluster_namespace="pingcap"
kubernetes_cluster_name="kind-cluster-1"
storageclass_name="local-storage"
```

执行以下指令，创建 `TidbMonitor`：

```sh
cat << EOF | kubectl apply -n ${cluster1_namespace} -f -
apiVersion: pingcap.com/v1alpha1
kind: TidbMonitor
metadata:
  name: ${cluster1_name}
spec:
  clusters:
  - name: ${cluster1_name}
    namespace: ${cluster1_namespace}
  externalLabels:

>     #kubernetes indicates the k8s cluster name, you can change the label's name on your own, but you should notice that `cluster` label has been used by tidb already. For more information, please refer to issue #4219.

Reviewer: ditto

    kubernetes: ${kubernetes_cluster1_name}
    #add other meta labels here
    #region: us-east-1
  initializer:
    baseImage: pingcap/tidb-monitor-initializer
    version: v5.2.1
  persistent: true
  storage: 20Gi
  storageClassName: ${storageclass_name}
  prometheus:
    baseImage: prom/prometheus
    logLevel: info
    version: v2.27.1
  reloader:
    baseImage: pingcap/tidb-monitor-reloader
    version: v1.0.1
  thanos:
    baseImage: quay.io/thanos/thanos
    version: v0.22.0
    #enable config below if long-term storage is needed.
    #objectStorageConfig:
    #  key: objectstorage.yaml
    #  name: thanos-objectstorage
  imagePullPolicy: IfNotPresent
```

</div>

<div label="Federation">

### 使用 Prometheus Federation

#### 部署架构图

> 使用 Federation Prometheus Server 作为数据统一存储与查询的入口，以笔者的经验来看，该方案只建议在数据规模较小的环境下使用。

Reviewer: `笔者` 看起来有点奇怪。我们官方文档应该是代表 PingCAP，而不只是文档 author？

---
title: 跨多个 Kubernetes 集群监控 TiDB 集群
summary: 介绍如何对跨多个 Kubernetes 集群的 TiDB 集群进行监控，并集成到常见 Prometheus 多集群监控体系中
---

# 跨多个 Kubernetes 集群监控 TiDB 集群

本文档介绍如何对跨多个 Kubernetes 集群的 TiDB 集群进行监控，以及如何与几种社区常见的 Prometheus 多集群监控方式进行集成，从而实现统一全局视图进行监控数据访问。

## Push 方式

### 部署架构图

> 本文档以 Thanos 为例，如果使用了其他兼容 Prometheus Remote API 的中心化存储方案（如 Cortex、M3DB、Grafana Cloud 等），只需对 Thanos 相关组件进行替换即可。

Author: 这里仅指 push

![push-thanos-receive.png](/media/push-thanos-receive.png)

### 前置条件

多个 Kubernetes 集群间的组件满足以下条件：

> - 各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件有能力访问 Thanos Receiver 组件。

Author: TidbMonitor 会启动 Prometheus Pod，二者可以理解为是同一个

remote_write_url="http://thanos-receiver:19291/api/v1/receive"
```

执行以下指令，创建 `TidbMonitor` ：

{{< copyable "shell-regular" >}}

```sh
cat << EOF | kubectl apply -n ${cluster_namespace} -f -
apiVersion: pingcap.com/v1alpha1
kind: TidbMonitor
metadata:
  name: ${cluster_name}
spec:
  clusters:
  - name: ${cluster_name}
    namespace: ${cluster_namespace}
  externalLabels:
    #kubernetes indicates the k8s cluster name, you can change the label's name on your own, but you should notice that `cluster` label has been used by tidb already. For more information, please refer to issue https://github.com/pingcap/tidb-operator/issues/4219.
    kubernetes: ${kubernetes_cluster_name}
    #add other meta labels here
    #region: us-east-1
  initializer:
    baseImage: pingcap/tidb-monitor-initializer
    version: v5.2.1
  persistent: true
  storage: 5Gi
  storageClassName: ${storageclass_name}
  prometheus:
    baseImage: prom/prometheus
    logLevel: info
    remoteWrite:
    - url: ${remote_write_url}
    retentionTime: 2h
    version: v2.27.1
  reloader:
    baseImage: pingcap/tidb-monitor-reloader
    version: v1.0.1
  imagePullPolicy: IfNotPresent
```

## Pull 方式

> 

Reviewer 3: 请在这里简要介绍什么是 pull 方式

---
title: 跨多个 Kubernetes 集群监控 TiDB 集群
summary: 介绍如何对跨多个 Kubernetes 集群的 TiDB 集群进行监控，并集成到常见 Prometheus 多集群监控体系中
---

# 跨多个 Kubernetes 集群监控 TiDB 集群

你可以监控跨多个 Kubernetes 集群的 TiDB 集群，实现从统一全局视图访问监控数据。本文档介绍如何与几种常见的 Prometheus 多集群监控方式进行集成：

- [Push 方式](#Push 方式)
- [Pull 方式 - 使用 Thanos Query](#使用 Thanos Query)
- [Pull 方式 - 使用 Prometheus Federation](#使用 Prometheus Federation)

## Push 方式

> > Push 方式利用 Prometheus remote-write 的特性，处在不同 Kubernetes 集群的 Prometheus 实例将监控数据推送至中心化存储中。

Reviewer 3: ```suggestion
Push 方式指利用 Prometheus remote-write 的特性，使位于不同 Kubernetes 集群的 Prometheus 实例将监控数据推送至中心化存储中。
```

### 部署架构图

> 本文档以 Thanos 为例，如果使用了其他兼容 Prometheus Remote API 的中心化存储方案（参考[Prometheus 集成方案](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage)），只需对 Thanos 相关组件进行替换即可。

Reviewer 3: ```suggestion
本节所描述的 Push 方式以 Thanos 为例。如果你使用了其他[兼容 Prometheus Remote API 的中心化存储方案](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage)，只需对 Thanos 相关组件进行替换即可。
```

![push-thanos-receive.png](/media/push-thanos-receive.png)

### 前置条件

多个 Kubernetes 集群间的组件满足以下条件：

> - 各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件有能力访问 Thanos Receiver 组件。

Reviewer 3: ```suggestion
- 各 Kubernetes 集群上的 Prometheus（即 TidbMonitor）组件有能力访问 Thanos Receiver 组件。
```

- Grafana 组件有能力访问 Thanos Query 组件。

### 部署 TiDB 集群监控

1. 根据不同 TiDB 集群所在的 Kubernetes 集群，设置以下环境变量：

    - `cluster_name`：TiDB 集群名称。
    - `cluster_namespace`：TiDB 集群所在的命名空间。
    - `kubernetes_cluster_name`：自定义的 Kubernetes 集群名称，在标识 Prometheus 的 `externallabels` 中使用。
    - `storageclass_name`：当前集群中的存储。
    - `remote_write_url`：`thanos-receiver` 组件的 host，或其他兼容 Prometheus remote API 组件的 host 。

>     如需部署 Thanos Receiver，可参考 [kube-thanos](https://github.com/thanos-io/kube-thanos) 以及 [Example](https://github.com/pingcap/tidb-operator/tree/master/examples/monitor-prom-remotewrite)。

Reviewer 3: 用户是否知道，什么情况下需要部署 Thanos Receiver？

    {{< copyable "shell-regular" >}}

    ```sh
    cluster_name="cluster1"
    cluster_namespace="pingcap"
    kubernetes_cluster_name="kind-cluster-1"
    storageclass_name="local-storage"
    remote_write_url="http://thanos-receiver:19291/api/v1/receive"
    ```

2. 执行以下指令，创建 `TidbMonitor`：

    {{< copyable "shell-regular" >}}

    ```sh
    cat << EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
    name: ${cluster_name}
    spec:
    clusters:
    - name: ${cluster_name}
        namespace: ${cluster_namespace}
    externalLabels:
        #kubernetes indicates the k8s cluster name, you can change the label's name on your own, but you should notice that `cluster` label has been used by tidb already. For more information, please refer to issue https://github.com/pingcap/tidb-operator/issues/4219.
        kubernetes: ${kubernetes_cluster_name}
        #add other meta labels here
        #region: us-east-1
    initializer:
        baseImage: pingcap/tidb-monitor-initializer

>         version: v5.2.1

Reviewer 3: 版本号需要更新吗？

    persistent: true
    storage: 5Gi
    storageClassName: ${storageclass_name}
    prometheus:
        baseImage: prom/prometheus
        logLevel: info
        remoteWrite:
        - url: ${remote_write_url}
        retentionTime: 2h
        version: v2.27.1
    reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
    imagePullPolicy: IfNotPresent
    ```

## Pull 方式

> > Pull 方式从不同 Kubernetes 集群的 Prometheus 实例中拉取监控数据，聚合后提供统一全局视图查询。

Reviewer 3: ```suggestion
Pull 方式是指从不同 Kubernetes 集群的 Prometheus 实例中拉取监控数据，聚合后提供统一全局视图查询。本文中提供两个示例：[使用 Thanos Query](#使用-thanos-query) 和 [使用 Prometheus Federation](使用-prometheus-federation)。
```

<SimpleTab>
<div label="Thanos Query">

### 使用 Thanos Query

#### 部署架构图

> 为每个 prometheus(TidbMonitor) 组件部署 thanos sidecar，并使用 thanos-query 组件进行聚合查询，其中 thanos-store、 S3 等组件在不需要对监控数据做长期存储时可以选择不部署。

Reviewer 3: ```suggestion
本节中的示例为每个 Prometheus (TidbMonitor) 组件部署了 Thanos Sidecar，并使用 thanos-query 组件进行聚合查询。如果不需要对监控数据做长期存储，你可以不部署 thanos-store、S3 等组件。
```

![pull-thanos-query.png](/media/pull-thanos-query.png)

#### 前置条件

需要配置 Kubernetes 的网络和 DNS，使得 Kubernetes 集群满足以下条件：

> - Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件的 Pod IP。
> - Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件的 Pod FQDN。

Reviewer 3: ```suggestion
- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus (TidbMonitor) 组件的 Pod IP。
- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus (TidbMonitor) 组件的 Pod FQDN。
```

        kubernetes: ${kubernetes_cluster_name}
        #add other meta labels here
        #region: us-east-1
    initializer:
        baseImage: pingcap/tidb-monitor-initializer
        version: v5.2.1
    persistent: true
    storage: 20Gi
    storageClassName: ${storageclass_name}
    prometheus:
        baseImage: prom/prometheus
        logLevel: info
        version: v2.27.1
    reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
    thanos:
        baseImage: quay.io/thanos/thanos
        version: v0.22.0
        #enable config below if long-term storage is needed.
        #objectStorageConfig:
        #  key: objectstorage.yaml
        #  name: thanos-objectstorage
    imagePullPolicy: IfNotPresent
    ```

</div>

<div label="Federation">

### 使用 Prometheus Federation

#### 部署架构图

使用 Federation Prometheus Server 作为数据统一存储与查询的入口，建议在数据规模较小的环境下使用。

![pull-prom-federation.png](/media/pull-prom-federation.png)

#### 前置条件

需要配置 Kubernetes 的网络和 DNS，使得 Kubernetes 集群满足以下条件：

> - Federation Prometheus 组件有能力访问各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件的 Pod IP。
> - Federation Prometheus 组件有能力访问各 Kubernetes 集群上的 Prometheus(TidbMonitor) 组件的 Pod FQDN。
> 

Reviewer 3: ```suggestion
- Federation Prometheus 组件有能力访问各 Kubernetes 集群上的 Prometheus (TidbMonitor) 组件的 Pod IP。
- Federation Prometheus 组件有能力访问各 Kubernetes 集群上的 Prometheus (TidbMonitor) 组件的 Pod FQDN。

```

    ```

2. 执行以下指令，创建 `TidbMonitor` ：

    {{< copyable "shell-regular" >}}

    ```sh
    cat << EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
    name: ${cluster_name}
    spec:
    clusters:
    - name: ${cluster_name}
        namespace: ${cluster_namespace}
    externalLabels:
        #kubernetes indicates the k8s cluster name, you can change the label's name on your own, but you should notice that `cluster` label has been used by tidb already. For more information, please refer to issue https://github.com/pingcap/tidb-operator/issues/4219.
        kubernetes: ${kubernetes_cluster_name}
        #add other meta labels here
        #region: us-east-1
    initializer:
        baseImage: pingcap/tidb-monitor-initializer
        version: v5.2.1
    persistent: true
    storage: 20Gi
    storageClassName: ${storageclass_name}
    prometheus:
        baseImage: prom/prometheus
        logLevel: info
        version: v2.27.1
    reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
    imagePullPolicy: IfNotPresent
    ```

#### 配置 Federation Prometheus

> 关于 Federation 方案，参考[federation文档](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation)。部署完成后修改 Prometheus 采集配置，添加需要汇总的 Prometheus(TiDBMonitor) 的 host 信息。

Reviewer 3: ```suggestion
关于 Federation 方案，参考 [Federation 文档](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation)。部署完成后，修改 Prometheus 采集配置，添加需要汇总的 Prometheus (TiDBMonitor) 的 host 信息。
```

        #add other meta labels here
        #region: us-east-1
      initializer:
        baseImage: pingcap/tidb-monitor-initializer
        version: v5.4.0
      persistent: true
      storage: 20Gi
      storageClassName: ${storageclass_name}
      prometheus:
        baseImage: prom/prometheus
        logLevel: info
        version: v2.27.1
      reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
      imagePullPolicy: IfNotPresent
    EOF
    ```

#### 配置 Federation Prometheus

关于 Federation 方案，参考 [Federation 文档](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation)。部署完成后，修改 Prometheus 采集配置，添加需要聚合的 Prometheus (TidbMonitor) 的 host 信息。

```
scrape_configs:
  - job_name: 'federate'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{__name__=~".+"}'

    static_configs:
      - targets:
        - 'source-prometheus-1:9090'
        - 'source-prometheus-2:9090'
        - 'source-prometheus-3:9090'
```

</div>
</SimpleTab>

## 使用 Grafana 可视化多集群监控数据

1. [配置 Prometheus 数据源](https://grafana.com/docs/grafana/latest/datasources/prometheus/)

> 2. 执行以下指令，获取 TiDB 相关组件的 Grafana Dashboards:

Reviewer 2: 这个放第一步吧，说一下 命令执行完以后，dashboards 在哪个目录，脚本里面有些环境变量怎么设置的？

Reviewer 2: 导出的 Dashboard 用户导入后，怎么在 Dashboard 上区分不同 k8s 的不同集群？

### 部署 TiDB 集群监控

1. 根据不同 TiDB 集群所在的 Kubernetes 集群，设置以下环境变量：

    - `cluster_name`：TiDB 集群名称。
    - `cluster_namespace`：TiDB 集群所在的命名空间。
    - `kubernetes_cluster_name`：自定义的 Kubernetes 集群名称，在标识 Prometheus 的 `externallabels` 中使用。
    - `storageclass_name`：当前集群中的存储。
    - `remote_write_url`：`thanos-receiver` 组件的 host，或其他兼容 Prometheus remote API 组件的 host 。

    关于 Thanos Receiver 部署，可参考 [kube-thanos](https://github.com/thanos-io/kube-thanos) 以及 [Example](https://github.com/pingcap/tidb-operator/tree/master/examples/monitor-prom-remotewrite)。

    {{< copyable "shell-regular" >}}

    ```sh
    cluster_name="cluster1"
    cluster_namespace="pingcap"
    kubernetes_cluster_name="kind-cluster-1"
    storageclass_name="local-storage"
    remote_write_url="http://thanos-receiver:19291/api/v1/receive"
    ```

2. 执行以下指令，创建 `TidbMonitor`：

    {{< copyable "shell-regular" >}}

    ```sh
    cat << EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
    name: ${cluster_name}
    spec:
    clusters:
    - name: ${cluster_name}
      namespace: ${cluster_namespace}
    externalLabels:

>       #kubernetes indicates the k8s cluster name, you can change the label's name on your own, but you should notice that `cluster` label has been used by tidb already. For more information, please refer to issue https://github.com/pingcap/tidb-operatorissues/4219.

Reviewer 2: ```suggestion
      # kubernetes indicates the k8s cluster name, you can change
      # the label's name on your own, but you should notice that the
      # `cluster` label has been used by the TiDB metrics already.
      # For more information, please refer to the issue
      # https://github.com/pingcap/tidb-operator/issues/4219.
```

      kubernetes: ${kubernetes_cluster_name}
      #add other meta labels here
      #region: us-east-1
    initializer:
      baseImage: pingcap/tidb-monitor-initializer
      version: v5.4.0
    persistent: true

>     storage: 5Gi

Reviewer 2: ```suggestion
    storage: 100Gi
```

    storageClassName: ${storageclass_name}
    prometheus:
      baseImage: prom/prometheus
      logLevel: info
      remoteWrite:
      - url: ${remote_write_url}

>       retentionTime: 2h

Reviewer 2: 这个可以先配置 7 天吧

Author: 这里 Push 的方式，数据不从 Local Prometheus 取，所以配置时间短一些

- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus (TidbMonitor) 组件的 Pod IP。
- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus (即 TidbMonitor) 组件的 Pod FQDN。

#### 部署 TiDB 集群监控

1. 根据不同 TiDB 集群所在的 Kubernetes 集群，设置以下环境变量：

    - `cluster_name`：TiDB 集群名称。
    - `cluster_namespace`：TiDB 集群所在的命名空间。
    - `kubernetes_cluster_name`：自定义的 Kubernetes 集群名称，在标识 Prometheus 的 `externallabels` 中使用。
    - `storageclass_name`：当前集群中的存储。

    关于 Thanos Query 部署, 参考 [kube-thanos](https://github.com/thanos-io/kube-thanos) 以及 [Example](https://github.com/pingcap/tidb-operator/tree/master/examples/monitor-with-thanos)。

    {{< copyable "shell-regular" >}}

    ```sh
    cluster_name="cluster1"
    cluster_namespace="pingcap"
    kubernetes_cluster_name="kind-cluster-1"
    storageclass_name="local-storage"
    ```

2. 执行以下指令，创建 `TidbMonitor`：

    {{< copyable "shell-regular" >}}

    ```sh
    cat << EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
    name: ${cluster_name}
    spec:
    clusters:
    - name: ${cluster_name}
      namespace: ${cluster_namespace}
    externalLabels:

>       #kubernetes indicates the k8s cluster name, you can change the label's name on your own, but you should notice that `cluster` label has been used by tidb already. For more information, please refer to issue https://github.com/pingcap/tidb-operator/issues/4219.

Reviewer 2: 同上

    storageClassName: ${storageclass_name}
    prometheus:
      baseImage: prom/prometheus
      logLevel: info
      version: v2.27.1
    reloader:
      baseImage: pingcap/tidb-monitor-reloader
      version: v1.0.1
    imagePullPolicy: IfNotPresent
    ```

#### 配置 Federation Prometheus

关于 Federation 方案，参考 [Federation 文档](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation)。部署完成后，修改 Prometheus 采集配置，添加需要聚合的 Prometheus (TidbMonitor) 的 host 信息。

```
scrape_configs:
  - job_name: 'federate'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{__name__=~".+"}'

    static_configs:
      - targets:
        - 'source-prometheus-1:9090'
        - 'source-prometheus-2:9090'
        - 'source-prometheus-3:9090'
```

</div>
</SimpleTab>

## 使用 Grafana 可视化多集群监控数据

1. [配置 Prometheus 数据源](https://grafana.com/docs/grafana/latest/datasources/prometheus/)

2. 执行以下指令，获取 TiDB 相关组件的 Grafana Dashboards:

    <SimpleTab>
    <div label="Linux">
    {{< copyable "shell-regular" >}}

    ```sh
    version=v5.4.0
    git clone -b auto-generate-for-$version https://github.com/pingcap/monitoring.git && \

>     cd monitoring/monitor-snapshot/$version/operator/dashboards && \

Reviewer 2: `monitor-snapshot` 不用了，现在是 build image 的时候临时去拉，这里直接用 Initializer image 起 container cp 出来吧

### 部署 TiDB 集群监控

1. 根据不同 TiDB 集群所在的 Kubernetes 集群，设置以下环境变量：

    - `cluster_name`：TiDB 集群名称。
    - `cluster_namespace`：TiDB 集群所在的命名空间。
    - `kubernetes_cluster_name`：自定义的 Kubernetes 集群名称，在标识 Prometheus 的 `externallabels` 中使用。
    - `storageclass_name`：当前集群中的存储。
    - `remote_write_url`：`thanos-receiver` 组件的 host，或其他兼容 Prometheus remote API 组件的 host 。

    关于 Thanos Receiver 部署，可参考 [kube-thanos](https://github.com/thanos-io/kube-thanos) 以及 [Example](https://github.com/pingcap/tidb-operator/tree/master/examples/monitor-prom-remotewrite)。

    {{< copyable "shell-regular" >}}

    ```sh
    cluster_name="cluster1"
    cluster_namespace="pingcap"
    kubernetes_cluster_name="kind-cluster-1"
    storageclass_name="local-storage"
    remote_write_url="http://thanos-receiver:19291/api/v1/receive"
    ```

2. 执行以下指令，创建 `TidbMonitor`：

    {{< copyable "shell-regular" >}}

    ```sh
    cat << EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
      name: ${cluster_name}
    spec:
      clusters:
      - name: ${cluster_name}
        namespace: ${cluster_namespace}
      externalLabels:

>         # kubernetes indicates the k8s cluster name, you can change

Reviewer: ```suggestion
        # k8s_clsuter indicates the k8s cluster name, you can change
```

        # the label's name on your own, but you should notice that the
        # "cluster" label has been used by the TiDB metrics already.
        # For more information, please refer to the issue
        # https://github.com/pingcap/tidb-operator/issues/4219.
        k8s_clsuter: ${kubernetes_cluster_name}

>         #add other meta labels here

Reviewer: ```suggestion
        # add other meta labels here
```

        #region: us-east-1
      initializer:
        baseImage: pingcap/tidb-monitor-initializer
        version: v5.4.0
      persistent: true
      storage: 100Gi
      storageClassName: ${storageclass_name}
      prometheus:
        baseImage: prom/prometheus
        logLevel: info
        remoteWrite:
        - url: ${remote_write_url}
        retentionTime: 2h
        version: v2.27.1
      reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
      imagePullPolicy: IfNotPresent
    EOF
    ```

## Pull 方式

Pull 方式是指从不同 Kubernetes 集群的 Prometheus 实例中拉取监控数据，聚合后提供统一全局视图查询。本文中将其分为：[使用 Thanos Query](#使用-thanos-query) 和 [使用 Prometheus Federation](#使用-prometheus-federation)。

<SimpleTab>
<div label="Thanos Query">

### 使用 Thanos Query

#### 部署架构图

本节中的示例为每个 Prometheus (TidbMonitor) 组件部署了 Thanos Sidecar，并使用 thanos-query 组件进行聚合查询。如果不需要对监控数据做长期存储，你可以不部署 thanos-store、S3 等组件。

![pull-thanos-query.png](/media/pull-thanos-query.png)

#### 前置条件

需要配置 Kubernetes 的网络和 DNS，使得 Kubernetes 集群满足以下条件：

> - Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus (TidbMonitor) 组件的 Pod IP。

Reviewer: ```suggestion
- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus (即 TidbMonitor) 组件的 Pod IP。
```

和下面一行对齐

- Thanos Query 组件有能力访问各 Kubernetes 集群上的 Prometheus (即 TidbMonitor) 组件的 Pod FQDN。

#### 部署 TiDB 集群监控

1. 根据不同 TiDB 集群所在的 Kubernetes 集群，设置以下环境变量：

    - `cluster_name`：TiDB 集群名称。
    - `cluster_namespace`：TiDB 集群所在的命名空间。
    - `kubernetes_cluster_name`：自定义的 Kubernetes 集群名称，在标识 Prometheus 的 `externallabels` 中使用。
    - `storageclass_name`：当前集群中的存储。

    关于 Thanos Query 部署, 参考 [kube-thanos](https://github.com/thanos-io/kube-thanos) 以及 [Example](https://github.com/pingcap/tidb-operator/tree/master/examples/monitor-with-thanos)。

    {{< copyable "shell-regular" >}}

    ```sh
    cluster_name="cluster1"
    cluster_namespace="pingcap"
    kubernetes_cluster_name="kind-cluster-1"
    storageclass_name="local-storage"
    ```

> 2. 执行以下指令，创建 `TidbMonitor`：

Reviewer: 要再加个步骤提一嘴怎么在 Thaos 那边配置吗

    {{< copyable "shell-regular" >}}

    ```sh
    cat <<EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
      name: ${cluster_name}
    spec:
      clusters:
      - name: ${cluster_name}
        namespace: ${cluster_namespace}
      externalLabels:

>         # kubernetes indicates the k8s cluster name, you can change

Reviewer: ```suggestion
        # k8s_clsuter indicates the k8s cluster name, you can change
```

使用 Federation Prometheus Server 作为数据统一存储与查询的入口，建议在数据规模较小的环境下使用。

![pull-prom-federation.png](/media/pull-prom-federation.png)

#### 前置条件

需要配置 Kubernetes 的网络和 DNS，使得 Kubernetes 集群满足以下条件：

- Federation Prometheus 组件有能力访问各 Kubernetes 集群上的 Prometheus (即 TidbMonitor) 组件的 Pod IP。
- Federation Prometheus 组件有能力访问各 Kubernetes 集群上的 Prometheus (即 TidbMonitor) 组件的 Pod FQDN。

#### 部署 TiDB 集群监控

1. 根据不同 TiDB 集群所在的 Kubernetes 集群，设置以下环境变量：

    - `cluster_name`：TiDB 集群名称。
    - `cluster_namespace`：TiDB 集群所在的命名空间。
    - `kubernetes_cluster_name`：自定义的 Kubernetes 集群名称，在标识 Prometheus 的 `externallabels` 中使用。
    - `storageclass_name`：当前集群中的存储。

    {{< copyable "shell-regular" >}}

    ```sh
    cluster_name="cluster1"
    cluster_namespace="pingcap"
    kubernetes_cluster_name="kind-cluster-1"
    storageclass_name="local-storage"
    ```

2. 执行以下指令，创建 `TidbMonitor` ：

    {{< copyable "shell-regular" >}}

    ```sh
    cat << EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
      name: ${cluster_name}
    spec:
      clusters:
      - name: ${cluster_name}
        namespace: ${cluster_namespace}
      externalLabels:

>         # kubernetes indicates the k8s cluster name, you can change

Reviewer: ```suggestion
        # k8s_clsuter indicates the k8s cluster name, you can change
```

        # the label's name on your own, but you should notice that the
        # "cluster" label has been used by the TiDB metrics already.
        # For more information, please refer to the issue
        # https://github.com/pingcap/tidb-operator/issues/4219.
        k8s_clsuter: ${kubernetes_cluster_name}

>         #add other meta labels here

Reviewer: ```suggestion
        # add other meta labels here
```

scrape_configs:
  - job_name: 'federate'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{__name__=~".+"}'

    static_configs:
      - targets:
        - 'source-prometheus-1:9090'
        - 'source-prometheus-2:9090'
        - 'source-prometheus-3:9090'
```

</div>
</SimpleTab>

## 使用 Grafana 可视化多集群监控数据

1. 执行以下指令，获取 TiDB 相关组件的 Grafana Dashboards, 其中 version 变量的值为 initializer 镜像版本，与 TiDB 版本对应关系如下，目前仅 nightly 版本镜像适用于**多集群**监控:

    |  TiDB 版本 | Initializer 镜像版本 |
    |:---|:---|
    | dev               | nightly                 |

    {{< copyable "shell-regular" >}}

    ```sh
    # set tidb version here
    version=nightly
    docker run --rm -i -v ${PWD}/dashboards:/dashboards/ pingcap/tidb-monitor-initializer:${version} && \
    cd dashboards
    ```

    执行上述命令后，可以在当前目录下查看所有组件 dashboard 的 json 定义文件。

2. [配置 Prometheus 数据源](https://grafana.com/docs/grafana/latest/datasources/prometheus/)，为了与上述获得的 dashboard json 文件保持一致，需将数据源 Name 字段值配置为 `tidb-cluster`，或执行以下指令对上述 dashboard json 文件中的数据源名称进行替换。

    {{< copyable "shell-regular" >}}

    ```sh
    # define your datasource name here.
    DS_NAME=tidb-cluster

>     sed -i 's/"datasource": "tidb-cluster"/"datasource": "$DS_NAME"/g' *.json

Reviewer: 没太看懂这行，前后都是 `"datasource": "tidb-cluster"`？

Author: 修改了一下描述

---
title: 跨多个 Kubernetes 集群监控 TiDB 集群
summary: 介绍如何对跨多个 Kubernetes 集群的 TiDB 集群进行监控，并集成到常见 Prometheus 多集群监控体系中
---

# 跨多个 Kubernetes 集群监控 TiDB 集群

> 你可以监控跨多个 Kubernetes 集群的 TiDB 集群，实现从统一全局视图访问监控数据。本文档介绍如何与几种常见的 Prometheus 多集群监控方式进行集成：

Reviewer 3: 这篇文档还讲了 Grafana 可视化，麻烦补充到文档开头

    {{< copyable "shell-regular" >}}

    ```sh
    cat << EOF | kubectl apply -n ${cluster_namespace} -f -
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
      name: ${cluster_name}
    spec:
      clusters:
      - name: ${cluster_name}
        namespace: ${cluster_namespace}
      externalLabels:
        # k8s_clsuter indicates the k8s cluster name, you can change
        # the label's name on your own, but you should notice that the
        # "cluster" label has been used by the TiDB metrics already.
        # For more information, please refer to the issue
        # https://github.com/pingcap/tidb-operator/issues/4219.
        k8s_clsuter: ${kubernetes_cluster_name}
        # add other meta labels here
        #region: us-east-1
      initializer:
        baseImage: pingcap/tidb-monitor-initializer
        version: v5.4.0
      persistent: true
      storage: 20Gi
      storageClassName: ${storageclass_name}
      prometheus:
        baseImage: prom/prometheus
        logLevel: info
        version: v2.27.1
      reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
      imagePullPolicy: IfNotPresent
    EOF
    ```

> #### 配置 Federation Prometheus

Reviewer 3: 1. Federation Prometheus 需要部署吗？有的话部署步骤或者参考链接写一下
2. “配置 Federation Prometheus” 和“部署 TiDB 集群监控”有先后顺序吗？

Author: 1. 需要部署，部署方式就是部署 Prometheus，让用户参考https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation 这个文档就可以，不需要在 tidb-operator 的 doc 里面再赘述了。
2. 二者没有先后顺序。

关于 Federation 方案，参考 [Federation 文档](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation)。部署完成后，修改 Prometheus 采集配置，添加需要聚合的 Prometheus (TidbMonitor) 的 host 信息。

```
scrape_configs:
  - job_name: 'federate'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{__name__=~".+"}'

    static_configs:
      - targets:
        - 'source-prometheus-1:9090'
        - 'source-prometheus-2:9090'
        - 'source-prometheus-3:9090'
```

</div>
</SimpleTab>

## 使用 Grafana 可视化多集群监控数据

> 

Reviewer 3: 这里加一个类似的描述：“使用 Prometheus 获取数据后，你可以使用 Grafana 可视化多集群监控数据……“

---
title: 跨多个 Kubernetes 集群监控 TiDB 集群
summary: 介绍如何对跨多个 Kubernetes 集群的 TiDB 集群进行监控，并集成到常见 Prometheus 多集群监控体系中
---

# 跨多个 Kubernetes 集群监控 TiDB 集群

你可以监控跨多个 Kubernetes 集群的 TiDB 集群，实现从统一全局视图访问监控数据。本文档介绍如何与几种常见的 Prometheus 多集群监控方式进行集成，并使用 Grafana 可视化多集群数据：

- [Push 方式](#push-方式)
- [Pull 方式 - 使用 Thanos Query](#使用-thanos-query)
- [Pull 方式 - 使用 Prometheus Federation](#使用-prometheus-federation)

> - [使用 Grafana 可视化多集群监控数据](#使用-Grafana-可视化多集群监控数据)

Reviewer 3: ```suggestion
- [使用 Grafana 可视化多集群监控数据](#使用-grafana-可视化多集群监控数据)
```

        k8s_clsuter: ${kubernetes_cluster_name}
        # add other meta labels here
        #region: us-east-1
      initializer:
        baseImage: pingcap/tidb-monitor-initializer
        version: v5.4.0
      persistent: true
      storage: 20Gi
      storageClassName: ${storageclass_name}
      prometheus:
        baseImage: prom/prometheus
        logLevel: info
        version: v2.27.1
      reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
      imagePullPolicy: IfNotPresent
    EOF
    ```

#### 配置 Federation Prometheus

关于 Federation 方案，参考 [Federation 文档](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation)。部署完成后，修改 Prometheus 采集配置，添加需要聚合的 Prometheus (TidbMonitor) 的 host 信息。

```
scrape_configs:
  - job_name: 'federate'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{__name__=~".+"}'

    static_configs:
      - targets:
        - 'source-prometheus-1:9090'
        - 'source-prometheus-2:9090'
        - 'source-prometheus-3:9090'
```

</div>
</SimpleTab>

## 使用 Grafana 可视化多集群监控数据

> 1. 执行以下指令，获取 TiDB 相关组件的 Grafana Dashboards, 其中 version 变量的值为 initializer 镜像版本，与 TiDB 版本对应关系如下，目前仅 nightly 版本镜像适用于**多集群**监控:
> 
>     |  TiDB 版本 | Initializer 镜像版本 |
>     |:---|:---|
>     | dev               | nightly                 |
> 

Reviewer 2: ```suggestion
1. 执行以下指令，获取 TiDB 相关组件的 Grafana Dashboards, 其中 version 变量的值为 initializer 镜像版本，应该和 TiDB 版本保持一致，但是目前仅 **nightly** 版本 Initializer 镜像适用于**多 Kubernetes 集群**监控。

```
